arroyo-worker/src/network_manager.rs

```diff
@@ -177,39 +217,60 @@ impl OutNetworkLink {
     }
 
     pub async fn add_receiver(&mut self, quad: Quad, rx: Receiver<QueueItem>) {
-        self.receivers.push((quad, rx));
+        self.receivers.push(NetworkReceiver {
+            quad,
+            rx,
+            dictionary_tracker: Arc::new(Mutex::new(DictionaryTracker::new(true))),
+        });
     }
 
     pub fn start(mut self) {
         tokio::spawn(async move {
             let mut sel = InQReader::new();
-            for (quad, mut rx) in self.receivers {
+            for NetworkReceiver {
+                quad,
+                mut rx,
+                mut dictionary_tracker,
+            } in self.receivers
+            {
                 let stream = async_stream::stream! {
                     while let Some(item) = rx.recv().await {
-                        yield (quad, item);
+                        yield (quad, dictionary_tracker.clone(), item);
                     }
                 };
                 sel.push(Box::pin(stream));
             }
             let mut flush_interval: Interval = interval(Duration::from_millis(100));
 
-            todo!("networking");
-            // loop {
-            //     select! {
-            //         Some(((quad, msg), s)) = sel.next() => {
-            //             let QueueItem::Bytes(data) = msg else {
-            //                 panic!("non-byte data in network queue")
-            //             };
-            //             let frame = Header::from_quad(quad, data.len());
-            //             frame.write(Pin::new(&mut self.stream)).await;
-            //             self.stream.write_all(&data).await.unwrap();
-            //             sel.push(s);
-            //         }
-            //         _ = flush_interval.tick() => {
-            //             self.stream.flush().await.unwrap();
-            //         }
-            //     }
-            // }
+            let write_options = IpcWriteOptions::default();
+
+            loop {
+                select! {
+                    Some(((quad, dictionary_tracker, msg), s)) = sel.next() => {
+                        match msg {
+                            ArrowMessage::Signal(signal) => {
+                                let data = bincode::encode_to_vec(&signal, config::standard()).unwrap();
+                                let header = Header::from_quad(quad, data.len(), MessageType::Signal);
+                                header.write(&mut Pin::new(&mut self.stream)).await;
+                                self.stream.write_all(&data).await.unwrap();
+                            }
+                            ArrowMessage::Data(data) => {
+                                let (_, encoded_message) = {
```

Comment:
I think it wouldn't be too hard to pass along the dictionaries emitted here in the future.

Reply:
My concern with dictionaries in the context of streaming is that there's no garbage collection. Over time, they would become unbounded in size without another protocol to manage GCing them.